modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1783 @@
+
+/** */
+public class IgniteSnapshotManager extends GridCacheSharedManagerAdapter implements PartitionsExchangeAware  {
+    /** File with delta pages suffix. */
+    public static final String DELTA_SUFFIX = ".delta";
+
+    /** File name template consists of delta pages. */
+    public static final String PART_DELTA_TEMPLATE = PART_FILE_TEMPLATE + DELTA_SUFFIX;
+
+    /** File name template for index delta pages. */
+    public static final String INDEX_DELTA_NAME = INDEX_FILE_NAME + DELTA_SUFFIX;
+
+    /** The reason of checkpoint start for needs of bakcup. */
+    public static final String SNAPSHOT_CP_REASON = "Wakeup for checkpoint to take snapshot [name=%s]";
+
+    /** Default working directory for snapshot temporary files. */
+    public static final String DFLT_LOCAL_SNAPSHOT_DIRECTORY = "snapshots";
+
+    /** Default snapshot directory for loading remote snapshots. */
+    public static final String DFLT_SNAPSHOT_WORK_DIRECTORY = "snp";
+
+    /** Prefix for snapshot threads. */
+    private static final String SNAPSHOT_RUNNER_THREAD_PREFIX = "snapshot-runner";
+
+    /** Total number of thread to perform local snapshot. */
+    private static final int SNAPSHOT_THEEAD_POOL_SIZE = 4;
+
+    /** Default snapshot topic to receive snapshots from remote node. */
+    private static final Object DFLT_RMT_SNAPSHOT_TOPIC = GridTopic.TOPIC_RMT_SNAPSHOT.topic("0");
+
+    /** Cache group id parameter name for a file transmission. */
+    private static final String SNP_GRP_ID_PARAM = "grpId";
+
+    /** Cache partition id parameter name for a file transmission. */
+    private static final String SNP_PART_ID_PARAM = "partId";
+
+    /** Cache local node directory path name (e.g. db/IgniteNode0). */
+    private static final String SNP_DB_NODE_PATH_PARAM = "dbNodePath";
+
+    /** Cache directory parameter name for a file transmission. */
+    private static final String SNP_CACHE_DIR_NAME_PARAM = "cacheDirName";
+
+    /** Snapshot parameter name for a file transmission. */
+    private static final String SNP_NAME_PARAM = "snpName";
+
+    /** Map of registered cache snapshot processes and their corresponding contexts. */
+    private final ConcurrentMap<String, LocalSnapshotContext> localSnpCtxs = new ConcurrentHashMap<>();
+
+    /** Map of requested snapshot from remote node. */
+    private final ConcurrentMap<T2<UUID, String>, SnapshotTransmission> reqSnps = new ConcurrentHashMap<>();
+
+    /** All registered page writers of all running snapshot processes. */
+    private final ConcurrentMap<GroupPartitionId, List<PageStoreSerialWriter>> partWriters = new ConcurrentHashMap<>();
+
+    /** Lock to protect the resources is used. */
+    private final GridBusyLock busyLock = new GridBusyLock();
+
+    /** Main snapshot directory to store files. */
+    private File localSnpDir;
+
+    /** Working directory for loaded snapshots from remote nodes. */
+    private File snpWorkDir;
```

Comment:
It's not clear what the difference between localSnpDir and snpWorkDir. JavaDoc states that it's used for snapshots from remote nodes, but in code it's also used by local snapshots.

Reply:
Renamed to tmp snapshot directory. Improved javadoc.